File: src/main/java/de/hdawg/rankinginfo/api/security/BearerTokenFilter.java
package de.hdawg.rankinginfo.api.security;

import java.io.IOException;
import java.util.Map;

import jakarta.servlet.FilterChain;
import jakarta.servlet.ServletException;
import jakarta.servlet.http.HttpServletRequest;
import jakarta.servlet.http.HttpServletResponse;

import org.springframework.http.MediaType;
import org.springframework.web.filter.OncePerRequestFilter;
import tools.jackson.databind.ObjectMapper;

public final class BearerTokenFilter extends OncePerRequestFilter {

  private final ApiTokenValidator tokenValidator;
  private final ObjectMapper objectMapper;

  public BearerTokenFilter(ApiTokenValidator tokenValidator, ObjectMapper objectMapper) {
    this.tokenValidator = tokenValidator;
    this.objectMapper = objectMapper;
  }

  @Override
  protected void doFilterInternal(
      HttpServletRequest request, HttpServletResponse response, FilterChain filterChain)
      throws ServletException, IOException {
    var authHeader = request.getHeader("Authorization");
    var token = authHeader != null ? authHeader.replaceFirst("^Bearer ", "") : null;
    if (token == null || token.isBlank() || !tokenValidator.isValid(token)) {
      response.setStatus(HttpServletResponse.SC_UNAUTHORIZED);
      response.setContentType(MediaType.APPLICATION_JSON_VALUE);
      objectMapper.writeValue(response.getWriter(), Map.of("error", "Unauthorized"));
      return;
    }
    filterChain.doFilter(request, response);
  }
}

File: src/main/kotlin/de/hdawg/rankinginfo/api/security/BearerTokenFilter.kt
package de.hdawg.rankinginfo.api.security

import com.fasterxml.jackson.databind.ObjectMapper
import jakarta.servlet.FilterChain
import jakarta.servlet.http.HttpServletRequest
import jakarta.servlet.http.HttpServletResponse
import org.springframework.http.MediaType
import org.springframework.web.filter.OncePerRequestFilter

/**
 * Servlet filter that enforces Bearer token authentication on API requests.
 *
 * Accepts tokens from two sources (checked in combination):
 * - `configuredTokens`: tokens provided via application configuration
 * - `API_BEARER_TOKEN` environment variable
 *
 * Returns HTTP 401 with a JSON error body when no valid token is present.
 */
class BearerTokenFilter(
    private val configuredTokens: List<String>,
    private val objectMapper: ObjectMapper,
) : OncePerRequestFilter() {
    private val envToken: String? = System.getenv("API_BEARER_TOKEN")

    override fun doFilterInternal(
        request: HttpServletRequest,
        response: HttpServletResponse,
        filterChain: FilterChain,
    ) {
        val token = request.getHeader("Authorization")?.removePrefix("Bearer ")
        if (token.isNullOrBlank() || !isValidToken(token)) {
            response.status = HttpServletResponse.SC_UNAUTHORIZED
            response.contentType = MediaType.APPLICATION_JSON_VALUE
            objectMapper.writeValue(response.writer, mapOf("error" to "Unauthorized"))
            return
        }
        filterChain.doFilter(request, response)
    }

    private fun isValidToken(token: String): Boolean {
        val validTokens =
            buildSet {
                addAll(configuredTokens.filter { it.isNotBlank() })
                envToken?.takeIf { it.isNotBlank() }?.let { add(it) }
            }
        return token in validTokens
    }
}
